Web Inspector: Show grid/flex overlays when highlighting elements

dcrousso opened the pull request:
#### 7cd36695ab33c2bc51653d1124f4b0ff2a6130ba
<pre>
Web Inspector: Show grid/flex overlays when highlighting elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=251937">https://bugs.webkit.org/show_bug.cgi?id=251937</a>

Reviewed by Patrick Angle.

This makes it so that the experience is the same between element selection and highlighting elements in Web Inspector (e.g. via the Elements Tab). Otherwise, it could be confusing for a developer to see grid/flex/etc. overlays when in element selection but not when highlighting the same element in Web Inspector.

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/inspector/agents/InspectorDOMAgent.h:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::willDestroyFrontendAndBackend):
(WebCore::InspectorDOMAgent::handleTouchEvent):
(WebCore::InspectorDOMAgent::highlightMousedOverNode):
(WebCore::InspectorDOMAgent::setSearchingForNode):
(WebCore::InspectorDOMAgent::gridOverlayConfigFromInspectorObject):
(WebCore::InspectorDOMAgent::flexOverlayConfigFromInspectorObject):
(WebCore::InspectorDOMAgent::highlightSelector):
(WebCore::InspectorDOMAgent::highlightNode):
(WebCore::InspectorDOMAgent::highlightNodeList):
Add optional parameters for `DOM.GridOverlayConfig` and `DOM.FlexOverlayConfig` to `DOM.highlightSelector`, `DOM.highlightNode`, and `DOM.highlightNodeList`.

* Source/WebCore/inspector/InspectorOverlay.h:
* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::isInNodeList): Added.
(WebCore::InspectorOverlay::paint):
(WebCore::InspectorOverlay::getHighlight):
(WebCore::InspectorOverlay::hideHighlight):
(WebCore::InspectorOverlay::highlightNodeList):
(WebCore::InspectorOverlay::highlightNode):
(WebCore::InspectorOverlay::shouldShowOverlay const):
(WebCore::InspectorOverlay::showHighlightGridOverlayForNode): Deleted.
(WebCore::InspectorOverlay::hideHighlightGridOverlay): Deleted.
(WebCore::InspectorOverlay::showHighlightFlexOverlayForNode): Deleted.
(WebCore::InspectorOverlay::hideHighlightFlexOverlay): Deleted.
Move the logic for `*Highlight{Grid,Flex}Overlay` to be part of `highlightNodeList` and `highlightNode`, since they all operate on the same `m_highlightNode` (and `m_highlightNodeList`) anyways.

* Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
(WI.DOMManager.buildHighlightConfigs): Renamed from `buildHighlightConfig`.
(WI.DOMManager.prototype.highlightDOMNodeList):
(WI.DOMManager.prototype.highlightSelector):
(WI.DOMManager.prototype.set inspectModeEnabled):
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode.prototype.highlight):
Pass all the grid/flex overlay options to all callers of `DOM.highlightSelector`, `DOM.highlightNode`, and `DOM.highlightNodeList`.
Drive-by: Unify all the places where these arguments are generated to avoid repeated code.

Canonical link: <a href="https://commits.webkit.org/260061@main">https://commits.webkit.org/260061@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dedcf634812cd19afeafd9adada692818decde3b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106838 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39630 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116021 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115583 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110741 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17368 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7076 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99025 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112606 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13154 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96155 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40737 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95038 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27806 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82473 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96357 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9030 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29161 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/95822 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/6970 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9605 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6159 "Found 1 new test failure: fast/text/bulgarian-system-language-shaping.html (failure)") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15197 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48707 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104580 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6946 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11132 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25903 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -466,76 +493,106 @@ void InspectorOverlay::paint(GraphicsContext& context)
 }
 
-void InspectorOverlay::highlightNode(Node* node, const InspectorOverlay::Highlight::Config& highlightConfig)
+void InspectorOverlay::highlightNode(Node* node, const InspectorOverlay::Highlight::Config& highlightConfig, const std::optional<Grid::Config>& gridOverlayConfig, const std::optional<Flex::Config>& flexOverlayConfig)
```
Aside: I wonder if we should make this a `WeakPtr<Node>`... The fact that we keep track of a raw pointer to a node seems less than ideal.

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
```diff
@@ -633,7 +633,29 @@ WI.DOMNode = class DOMNode extends WI.Object
+                flexColor: WI.DOMNode.defaultLayoutOverlayColor.toProtocol(),
+                showOrderNumbers: WI.settings.flexOverlayShowOrderNumbers.value,
+            };
+        }
```
This code to build the various overlay configs from settings occurs three times in the patch, and I think at least once more already. Can we turn these in to a static helper somewhere that at least builds the protocol config payloads? This is a lot of call sites to go update if we ever add another parameter to an overlay's config.

dcrousso committed:
Web Inspector: Show grid/flex overlays when highlighting elements
https://bugs.webkit.org/show_bug.cgi?id=251937

Reviewed by Patrick Angle.

This makes it so that the experience is the same between element selection and highlighting elements in Web Inspector (e.g. via the Elements Tab). Otherwise, it could be confusing for a developer to see grid/flex/etc. overlays when in element selection but not when highlighting the same element in Web Inspector.

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WebCore/inspector/agents/InspectorDOMAgent.h:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::willDestroyFrontendAndBackend):
(WebCore::InspectorDOMAgent::handleTouchEvent):
(WebCore::InspectorDOMAgent::highlightMousedOverNode):
(WebCore::InspectorDOMAgent::setSearchingForNode):
(WebCore::InspectorDOMAgent::gridOverlayConfigFromInspectorObject):
(WebCore::InspectorDOMAgent::flexOverlayConfigFromInspectorObject):
(WebCore::InspectorDOMAgent::highlightSelector):
(WebCore::InspectorDOMAgent::highlightNode):
(WebCore::InspectorDOMAgent::highlightNodeList):
Add optional parameters for `DOM.GridOverlayConfig` and `DOM.FlexOverlayConfig` to `DOM.highlightSelector`, `DOM.highlightNode`, and `DOM.highlightNodeList`.

* Source/WebCore/inspector/InspectorOverlay.h:
* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::isInNodeList): Added.
(WebCore::InspectorOverlay::paint):
(WebCore::InspectorOverlay::getHighlight):
(WebCore::InspectorOverlay::hideHighlight):
(WebCore::InspectorOverlay::highlightNodeList):
(WebCore::InspectorOverlay::highlightNode):
(WebCore::InspectorOverlay::shouldShowOverlay const):
(WebCore::InspectorOverlay::showHighlightGridOverlayForNode): Deleted.
(WebCore::InspectorOverlay::hideHighlightGridOverlay): Deleted.
(WebCore::InspectorOverlay::showHighlightFlexOverlayForNode): Deleted.
(WebCore::InspectorOverlay::hideHighlightFlexOverlay): Deleted.
Move the logic for `*Highlight{Grid,Flex}Overlay` to be part of `highlightNodeList` and `highlightNode`, since they all operate on the same `m_highlightNode` (and `m_highlightNodeList`) anyways.

* Source/WebInspectorUI/UserInterface/Controllers/DOMManager.js:
(WI.DOMManager.buildHighlightConfigs): Renamed from `buildHighlightConfig`.
(WI.DOMManager.prototype.highlightDOMNodeList):
(WI.DOMManager.prototype.highlightSelector):
(WI.DOMManager.prototype.set inspectModeEnabled):
* Source/WebInspectorUI/UserInterface/Models/DOMNode.js:
(WI.DOMNode.prototype.highlight):
Pass all the grid/flex overlay options to all callers of `DOM.highlightSelector`, `DOM.highlightNode`, and `DOM.highlightNodeList`.
Drive-by: Unify all the places where these arguments are generated to avoid repeated code.

Canonical link: https://commits.webkit.org/260061@main
